source/common/rds/route_config_provider_manager.h

```diff
@@ -34,12 +35,43 @@ class RouteConfigProviderManager {
 
   RouteConfigProviderPtr
   addStaticProvider(std::function<RouteConfigProviderPtr()> create_static_provider);
+
+  template <class RdsConfig>
   RouteConfigProviderSharedPtr
-  addDynamicProvider(const Protobuf::Message& rds, const std::string& route_config_name,
+  addDynamicProvider(const RdsConfig& rds, const std::string& route_config_name,
```

Comment:
nit: it may be better to keep the method implementation in the .cc file.
If linking is an issue, it may be solved by explicitly stating the instances that are supported ([an example](https://github.com/envoyproxy/envoy/blob/e456b9ba7683a09cea7a7f2af6d3a5a5c2ac10a6/source/extensions/config_subscription/grpc/xds_mux/grpc_mux_impl.cc#L451-L456) that does this for templated classes, I'm assuming that templated functions are similar).

Reply:
@adisuissa I tried to move the templated method implementation back to .cc file, but run into some problems. 
In order to do explicit function template instantiation, the template definition must be visible from the place when explicit template instantiation happens. Since the template definition is moved back to source/common/rds/route_config_provider_manager.cc file, the explicit function instantiation must happen at the same file. It means that I would have to import protos of all rds configs which I know use addDynamicProvider (router, thrift, dubbo and maybe few more from contrib - did not check). To be honest I am not sure if this makes sense. Notice that in [the example](https://github.com/envoyproxy/envoy/blob/e456b9ba7683a09cea7a7f2af6d3a5a5c2ac10a6/source/extensions/config_subscription/grpc/xds_mux/grpc_mux_impl.cc#L451-L456), [method template definition](https://github.com/envoyproxy/envoy/blob/e456b9ba7683a09cea7a7f2af6d3a5a5c2ac10a6/source/extensions/config_subscription/grpc/xds_mux/grpc_mux_impl.cc#L433-L449) happens just before explicit instantiation.